#+TITLE: CollecTF documentation
#+AUTHOR: Sefa Kilic
#+email: sefa1@umbc.edu

CollecTF (http://collectf.umbc.edu), a play on words using the French collectif
[collective] and the acronym for transcription factor [TF], is a database of
transcription factor binding sites (TFBS) in the Bacteria domain.

CollecTF aims at becoming a reference, highly-accessed database by relying on its
ability to customize navigation and data extraction, its relevance to the community,
the quality and detail of the stored data and the up-to-date nature of the stored
information.

This documentation provides information about the database design and implementation
details. For more information about the database itself, visit
http://erilllab.biosci.umbc.edu/collectf/main_page_about/

* Database structure

Graphical representation of the database is available at :download:`collectf_db.pdf`.

The database consists of several tables and relationships between them. It is
implemented in MySQL with Django interface. Each table is represented as a Python
class (model) and access to the database is through instances of those classes.

Below, each table is described in detail.

- Curation :: adsad
- Curator :: dasd
- Publication :: 
- Gene ::
- Genome ::
- GenomeSequence :: 
- Taxonomy :: 
- TF ::
- TFFamily :: 
- TFInstance :: 
- SiteInstance ::
- Curation_SiteInstance ::
- Regulation ::
- NotAnnotatedSiteInstance ::
- ExperimentalTechnique ::
- ChipInfo ::
- ExternalDatabase ::
- Curation_ExternalDatabase ::
- NCBISubmission :: asdsa

* Curation process

** Publication submission
Publication submission is fairly simple. User puts the PMID for the paper. After
confirmation page, if the paper is not already in the database, it is added. The
paper can also be assigned to the curator.

** Curation submission
Curation submission consists of several steps. For each step, a form is displayed,
the curator fills all fields properly, submits the form and proceeds to the next
step.

** Publication selection
All papers assigned to the curator are listed. The curator selects the paper that
he/she is curating and moves to the next step.

** Accession numbers and general information
This step contains TF and genome information reported in the paper.

** Methodology
Experimental techniques used in the paper are submitted here.

** Site Entry
The reported sites are entered in this step.

** Exact site search
Sites that are entered in the previous step are searched in the genome sequence. For
a given site, there may be more than one exact matches. For each site that has at
least one exact match are presented in this step for curator to match reported sites
and matched instances in the genome.

** Soft site search
For some papers, the reported genome is not available in RefSeq database, so the
curator picks the closest genome from NCBI RefSeq database. In that case, for some
reported sites, the exact match may not be present in the surrogate genome.
   
For all reported sites that are not matched to any site instance in the previous step
(exact site search), a soft search is performed to find site instances that are not
exactly same but similar to the reported site (upto 2 mismatches).

** Gene regulation
** Curation review

* Operon Prediction
When the curator enters list of sites reported in the paper, all sites are searched
in the genome and all matches are returned. For each site-match in the genome, in
addition to the genomic position of the match, nearby genes are identified so that
they can be marked as regulated by TF (if any experimental evidence in the paper).

For each identified site instance, operons in both directions (if any) are
returned. The implementation for operon prediction is given below.

* External Submission
===================
Curation submission form for external users is same as default form except few
changes in the last step (curation review step). For external users, the field
``revision_reasons`` is hidden and set to ``external_submission`` and the field
``NCBI_submission_ready`` is hidden and set to false.

Currently, user registration is disabled and is handled via feedback form (user requests
an account and we open a new account through admin page).

* Browsing Database
** Browse by TF/species/techniques
** Search
** Compare motifsg

* NCBI Annotation
** Meta-site Rule
Experimental evidence for a given binding site may be reported in more than one
paper. In addition to that, in some cases, the coordinates for the same binding
location may be slightly different (a few bp). To avoid listing same binding
locations multiple times and reporting multiple annotations to NCBI RefSeq for the
same site instance), we introduce meta-site concept. A meta-site is the collection of
all evidences for the same genomic binding location, distributed across multiple
reported sites that are a few bp off each other. Each meta-site is represented by one
of its members (leader site).

Evidence from two motif-associated sites is combined into one meta-site if the
overlap between two sites is larger than 75% of the combined site length.  A
non-motif associated site is integrated into a meta-site if it fully overlaps with
any motif-associated site in the meta-site.

Meta-sites are not stored in the database and generated dynamically. For NCBI
annotation, overlapping site instances are merged into a meta-site and leader-site id
is used for dbxref link-out.

** dbxref ID generation
Each CollecTF annotation in NCBI RefSeq records has unique dbxref out-link to
CollecTF record, pointing a curated site instance. There is one-to-one mapping
between each annotated site-instance id and corresponding dbxref.

