#+TITLE: CollecTF Documentation
#+AUTHOR: Sefa Kilic


CollecTF (http://collectf.umbc.edu), a play on words using the French collectif
[collective] and the acronym for transcription factor [TF], is a database of
transcription factor binding sites (TFBS) in the Bacteria domain.

CollecTF aims at becoming a reference, highly-accessed database by relying on its
ability to customize navigation and data extraction, its relevance to the community,
the quality and detail of the stored data and the up-to-date nature of the stored
information.

This documentation provides information about the database design and implementation
details. For more information about the database itself, visit
http://collectf.umbc.edu


* Database structure
Graphical representation of the database is available at [[file:collectf_db.pdf][here]].

The database consists of several tables and relationships between them. It is
implemented using MySQL with Django interface. Each table is represented as a Python
class (model) and access to the database entries is performed through instances of
those classes.

Below, each table is described in detail.
- Curation :: 
- Curator :: 
- Publication :: 
- Gene ::
- Genome ::
- GenomeSequence :: 
- Taxonomy :: 
- TF ::
- TFFamily :: 
- TFInstance :: 
- SiteInstance ::
- Curation\_SiteInstance ::
- Regulation ::
- NotAnnotatedSiteInstance ::
- ExperimentalTechnique ::
- ChipInfo ::
- ExternalDatabase ::
- Curation\_ExternalDatabase ::
- NCBISubmission :: 

* Curation process
** Publication submission
Publication submission is fairly simple. User puts the PMID for the paper. After
confirmation page, if the paper is not already in the database, it is added. The user
can also assign the paper to himself/herself.

** Curation submission
Curation submission consists of several steps. For each step, a form is displayed,
the curator fills all fields properly, submits the form and proceeds to the next
step.

*** Publication selection
*** Accession numbers and general information
*** Methodology
*** Site entry
*** Exact site search
*** Soft site search
*** Gene regulation
*** Curation review

** External submission
Curation submission form for external users is same as default form except few
changes in the last step (curation review step). For external users, the field
=revision_reasons= is hidden and set to =external_submission= and the field
=NCBI_submission_ready= is hidden and set to false.


* Browsing database
** Browse by TF, species and experimental techniques
** Search
** Compare motifs

* Database statistics
* NCBI export
* Master-curator view

* Implementation notes
** Operon prediction
** Meta-site rule
Experimental evidence for a given binding site may be reported in more than one
paper. In addition to that, in some cases, the coordinates for the same binding
ocation may be slightly different (a few bp). To avoid listing same binding ocations
multiple times and reporting multiple annotations to NCBI RefSeq for the same site
instance), we introduce meta-site concept. A meta-site is the collection of all
evidences for the same genomic binding location, distributed across multiple reported
sites that are a few bp off each other. Each meta-site is represented by one of its
members (leader site).

Evidence from two motif-associated sites is combined into one meta-site if the
overlap between two sites is larger than 75% of the combined site length.  A
non-motif associated site is integrated into a meta-site if it fully overlaps with
any motif-associated site in the meta-site.

Meta-sites are not stored in the database and generated dynamically. For NCBI
annotation, overlapping site instances are merged into a meta-site and leader-site id
is used for dbxref link-out.

** dbxref id generation
Each CollecTF annotation in NCBI RefSeq records has unique dbxref out-link to
CollecTF record, pointing a curated site instance. There is one-to-one mapping
between each annotated site-instance id and corresponding dbxref.
